scipy/stats/_hypotests.py

```diff
@@ -663,3 +666,366 @@ def somersd(x, y=None):
+        A result object with the following attributes.
+
+        statistic : float
+            The Wald statistic with pooled or unpooled variance, depending
+            on the user choice of `pooled`.
+
+        pvalue : float
+            The probability of obtaining a test statistic at least as extreme
+            as the one observed under the null hypothesis.
+
+    See Also
+    --------
+    chi2_contingency : Chi-square test of independence of variables in a
+        contingency table.
+    fisher_exact : Fisher exact test on a 2x2 contingency table.
+
+    Notes
+    -----
+    Barnard's test is an exact test used in the analysis of contingency
+    tables. It examines the association of two categorical variables, and
+    is a more powerful alternative than Fisher's exact test
+    for 2x2 contingency tables.
+
+    Let's define :math:`X` a 2x2 matrix representing the contingency table,
+    where each column stores the binomial experiment, as in the example
+    below. Let's also define :math:`p_1, p_2` the theoretical binomial
+    probabilities for  :math:`x_{11}` and :math:`x_{12}`. When using
+    Barnard exact test, we can assert three different null hypotheses :
+
+    - :math:`H_0 : p_1 \geq p_2` versus :math:`H_1 : p_1 < p_2`,
+      with `alternative` = "less"
+
+    - :math:`H_0 : p_1 \leq p_2` versus :math:`H_1 : p_1 > p_2`,
+      with `alternative` = "greater"
+
+    - :math:`H_0 : p_1 = p_2` versus :math:`H_1 : p_1 \neq p_2`,
+      with `alternative` = "two-sided" (default one)
+
+    In order to compute Barnard's exact test, we are using the Wald
+    statistic [3]_ with pooled or unpooled variance.
+    Under the default assumption that both variances are equal
+    (``pooled = True``), the statistic is computed as:
+
+    .. math::
+
+        T(X) = \frac{
+            \hat{p}_1 - \hat{p}_2
+        }{
+            \sqrt{
+                \hat{p}(1 - \hat{p})
+                (\frac{1}{c_1} +
+                \frac{1}{c_2})
+            }
+        }
+
+    with :math:`\hat{p}_1, \hat{p}_2` and :math:`\hat{p}` the estimator of
+    :math:`p_1, p_2` and :math:`p`, the latter being the combined probability,
+    given the assumption that :math:`p_1 = p_2`.
+
+    If this assumption is invalid (``pooled = False``), the statistic is:
+
+    .. math::
+
+        T(X) = \frac{
+            \hat{p}_1 - \hat{p}_2
+        }{
+            \sqrt{
+                \frac{\hat{p}_1 (1 - \hat{p}_1)}{c_1} +
+                \frac{\hat{p}_2 (1 - \hat{p}_2)}{c_2}
+            }
+        }
+
+    The p-value is then computed as:
+
+    .. math::
+
+        pvalue =
+        \sum
+            \binom{c_1}{x_{11}}
+            \binom{c_2}{x_{12}}
+            \pi^{x_{11} + x_{12}}
+            (1 - \pi)^{n - x_{11} - x_{12}}
+
+    where :math:`T(X) \leq T(X_0)`, if `alternative` = "less"
+    :math:`T(X) \geq T(X_0)`, if `alternative` = "greater"
+    :math:`T(X) \geq |T(X_0)|`, if `alternative` = "two-sided".
```

Comment:
How does the alternative affect the formula for the p-value above? (If `max_{\pi \in [0, 1]}` is prepended to the formula, it looks like it can be computed without considering the alternative.)

What is the difference between X and X_0?

Reply:
Without considering the alternative: the sums is made of all X s.t. the stats is worse than the observed. Worse def depends to the alternative.

X is a contingency table (the test is made by enumerating all contingency table), X₀ is the observed contingency table.

Reply:
I figured. But at the top, 
```
Let's define :math:`X` a 2x2 matrix
```

So X_0 down here is the same as X above, and X down here isn't defined. One way to fix this would be 
```
"Let's define :math:`X_0` a 2x2 matrix"
```
above.

Reply:
I approved, but this is still unresolved.